### PR TITLE
Path patterns support

### PR DESCRIPTION
**Pull request by gsvgit**
Hi @cleishm 

The syntax for path patterns as described in [this proposal](https://github.com/thobe/openCypher/blob/rpq/cip/1.accepted/CIP2017-02-06-Path-Patterns.adoc) (with [this fix](https://github.com/opencypher/openCypher/pull/187#discussion_r406155831)) is supported. This extension allows one to express RPQs and CFPQs (regular and context-free path queries) in Cypher.
- [Syntax description.](https://github.com/thobe/openCypher/blob/rpq/cip/1.accepted/CIP2017-02-06-Path-Patterns.adoc#syntax)
-  [Discussion](https://github.com/opencypher/openCypher/pull/187) of the proposal in the openCypher community.

### CODE REVIEW

**Comment by cleishm** [comment]
Wow. This is amazing.

I'll take a look through!

**Comment by cleishm** [comment]
Quick question while I'm reviewing. I'm unfamiliar with the latest CIP, so perhaps I'm not following. From the CIP, it seems that "path base" can never appear as an AST node itself:

```
PathBase = PathEdge
         | PathAny
         | PathNode
         | PathReference
         | PathGroup
         ;
```

But this PR does add it as a concrete AST node. If there is common functionality (likely), then perhaps it could be a sub-type instead? Examples of this in the current codebase are `cypher_expression_astnode_vt` and `cypher_query_option_astnode_vt`.

**Comment by simpletonDL** [comment]
Hi, Chris
Thanks for the detailed question. This is probably not a very suitable name for this concrete AST node. In my mind PathBase is union of PathRepetition and PathDirection, so it stores direction and range details. Something like this:
```
PathBase = ['<'], SomeNode, ['>'], [('*', [RangeDetail]) | '+' | '?'];
SomeNode = PathEdge
         | PathAny
         | PathNode
         | PathReference
         | PathGroup
         ;
```
I maybe don`t understand what exactly sub-type is, but it doesn't look like a something that can contains useful fields. So I believe that this is not what we want.

File: lib/src/parser.leg [hunk 1]
```diff
@@ -84,6 +84,11 @@ schema-command =
     ) ~{ERR("a schema command")}
     (SEMICOLON | EOF)
 
+named-path-clause =
+    < PATHPATTERN i:identifier EQUAL - p:pattern-path -
+    (WHERE e:expression | e:_null_) 
+    >                                   { $$ = named_path_predicate(i, p, e); }
```
Comment (simpletonDL) [inline lib/src/parser.leg:90]: Here we use named_path_predicate (almost the same as named_path) for named path patterns construction. So the node of type CYPHER_AST_NAMED_PATH may appear both as child of pattern and as child of query. Named path patterns and named paths are structurally identical but different in semantics. So should we create new concrete AST node for named-path-clause in this case?

**Comment by gsvgit** [comment]
Hi @cleishm! 
For what reason this pull request was closed?

**Comment by cleishm** [comment]
@gsvgit Ah - not intentional. Please feel free to re-open against the main branch!

**Pull request closed without merging**